File: templates/apps/api-python/src/[zitadel]stream.py
import asyncio
import json
from redis.asyncio import Redis
from .models.user import User

_STREAM_KEY = "str:zitadel:events"
_CONSUMER_GROUP = "group:{{snakeCase name}}-service"


async def _process_event(event_type: str, data: dict) -> None:
    if event_type in ("user.registered", "user.updated"):
        await User.update_or_create(
            defaults={"email": data["email"], "name": data.get("name")},
            zitadel_id=data["zitadel_id"],
        )
        print(f"[Sync] Processed {event_type} for user: {data['email']}")


async def setup_consumer(redis: Redis) -> None:
    consumer_name = f"consumer:{id(asyncio.current_task())}"

    try:
        await redis.xgroup_create(_STREAM_KEY, _CONSUMER_GROUP, id="0", mkstream=True)
        print(f"[Redis] Consumer group {_CONSUMER_GROUP} initialized.")
    except Exception as e:
        if "BUSYGROUP" not in str(e):
            raise

    print(f"[Redis] {consumer_name} listening for identity events...")

    while True:
        try:
            result = await redis.xreadgroup(
                groupname=_CONSUMER_GROUP,
                consumername=consumer_name,
                streams={_STREAM_KEY: ">"},
                block=5000,
                count=10,
            )
            if result:
                for _, messages in result:
                    for msg_id, fields in messages:
                        event_type = fields.get("type", "")
                        payload_str = fields.get("payload", "{}")
                        data = json.loads(payload_str)
                        await _process_event(event_type, data)
                        await redis.xack(_STREAM_KEY, _CONSUMER_GROUP, msg_id)
        except asyncio.CancelledError:
            break
        except Exception as e:
            print(f"[Redis] Stream reading error: {e}")
            await asyncio.sleep(2)

File: templates/apps/api/src/[zitadel]stream.ts
import Redis from 'ioredis';
import { db, users } from '@workspace/db-{{dashCase name}}';

const redisUrl = process.env.REDIS_URL || 'redis://localhost:6379/1';
const streamKey = 'str:zitadel:events';
const consumerGroup = 'group:{{dashCase name}}-service';
const consumerName = `consumer:${Math.random().toString(36).substring(7)}`;

// We export a singleton redis client for publishing (e.g. from Platform API)
export const redis = new Redis(redisUrl);

export async function setupConsumer() {
  const consumeClient = new Redis(redisUrl);

  // Ensure consumer group exists
  try {
    await consumeClient.xgroup('CREATE', streamKey, consumerGroup, '0', 'MKSTREAM');
    console.log(`[Redis] Consumer group ${consumerGroup} initialized.`);
  } catch (err: any) {
    if (!err.message.includes('BUSYGROUP')) {
      console.error('[Redis] Failed to create consumer group', err);
    }
  }

  console.log(`[Redis] ${consumerName} listening for identity events...`);

  // Infinite polling loop for XREADGROUP
  while (true) {
    try {
      // Block for 5000ms waiting for new messages
      const result = await consumeClient.xreadgroup(
        'GROUP', consumerGroup, consumerName,
        'BLOCK', 5000,
        'STREAMS', streamKey, '>'
      );

      if (result) {
        const stream = result[0];
        const messages = stream[1];

        for (const message of messages) {
          const [messageId, fields] = message;

          let type = '';
          let payloadStr = '';
          for (let i = 0; i < fields.length; i += 2) {
            if (fields[i] === 'type') type = fields[i+1];
            if (fields[i] === 'payload') payloadStr = fields[i+1];
          }

          if (payloadStr) {
            const data = JSON.parse(payloadStr);
            await processIdentityEvent(type, data);
          }

          // Acknowledge processing
          await consumeClient.xack(streamKey, consumerGroup, messageId);
        }
      }
    } catch (err) {
      console.error('[Redis] Stream reading error:', err);
      // Wait before retrying to prevent hot looping on failure
      await new Promise(res => setTimeout(res, 2000));
    }
  }
}

async function processIdentityEvent(type: string, data: any) {
  try {
    if (type === 'user.registered' || type === 'user.updated') {
      // Upsert logic to handle both creation and updates
      await db.insert(users).values({
        zitadelId: data.zitadelId,
        email: data.email,
        name: data.name,
      }).onConflictDoUpdate({
        target: users.zitadelId,
        set: { email: data.email, name: data.name, updatedAt: new Date() }
      });
      console.log(`[Sync] Processed ${type} for user: ${data.email}`);
    }
  } catch (error) {
    console.error(`[Sync] Failed to process ${type}:`, error);
  }
}
